runtime/interface.go

```diff
@@ -34,36 +34,25 @@ type Interface interface {
 	ResolveLocation(identifiers []Identifier, location Location) ([]ResolvedLocation, error)
 	// GetCode returns the code at a given location
 	GetCode(location Location) ([]byte, error)
-	// GetProgram returns the program for the given location, if available.
-	//
-	// NOTE:
+	// GetAndSetProgram returns the program for the given location, if available,
+	// or sets the program by calling the given load function.
 	//
 	// For implementations:
-	// - During execution, this function MUST always return the *same* program,
+	// - Perform a lookup for the location and return the program if it exists.
+	// - If the program does not exist, call load, and store the result,
+	//   *EVEN IF loading failed* (program is nil / error is non-nil)!
+	// - During execution of a high-level program (e.g. script, transaction, etc.),
+	//   this function MUST always return the *same* program,
 	//   i.e. it may NOT return a different program,
 	//   an elaboration in the program that is not annotating the AST in the program;
 	//   or a program/elaboration and then nothing in a subsequent call.
-	// - This function MUST also return what was set using SetProgram,
-	//   it may NOT return something different or nothing/nil (!) after SetProgram was called.
-	//   Do NOT implement this as a cache!
-	//
-	// For uses:
-	// - ONLY call this function when a program must be parsed and checked,
-	//   as an optimization to reuse the result of a potential previous parse and check.
-	// - If GetProgram returns nil, Cadence MUST call SetProgram:
-	//   There's an informal contract between Cadence and the implementer:
-	//   Cadence calls GetProgram to potentially avoid having to parse and check a program.
-	//   If the implementer returns nil from GetProgram,
-	//   it expects that Cadence sets the resulting parsed and checked program with SetProgram.
-	// - The behaviour after GetProgram returning nil or a program must be always deterministic:
-	//   As SetProgram is called when GetProgram is nil, then SetProgram MUST also be called when
-	//   GetProgram returns a program. This prevents nondeterministic behaviour
-	//
-	// Deprecated: This function should be refactored to ensure that SetProgram is always called
-	//
-	GetProgram(Location) (*interpreter.Program, error)
-	// SetProgram sets the program for the given location.
-	SetProgram(Location, *interpreter.Program) error
+	// - This function MUST also return what was previously loaded,
+	//   it may NOT return something different or nothing/nil (!) after load was called.
```

Comment:
what about the case where load itself returns:
- nil
- or an error

Reply:
Tried to improve the comment further in [e3cfb8e](https://github.com/onflow/cadence/pull/2254/commits/e3cfb8eee8c43a54d5ae2c8eb400e650324fe9a9)